src/boot/chid.h

```diff
@@ -45,20 +55,38 @@ assert_cc(offsetof(Device, descriptor) == 0);
 assert_cc(offsetof(Device, chid) == 4);
 assert_cc(offsetof(Device, devicetree.name_offset) == 20);
 assert_cc(offsetof(Device, devicetree.compatible_offset) == 24);
+assert_cc(offsetof(Device, fw.name_offset) == 20);
+assert_cc(offsetof(Device, fw.compatible_offset) == 24);
 assert_cc(sizeof(Device) == 28);
 
 static inline const char* device_get_name(const void *base, const Device *device) {
-        if (device->descriptor != DEVICE_DESCRIPTOR_DEVICETREE)
+        uint32_t off;
```

Comment:
Needs to be zero-initialized

Reply:
Why? Its getting assigned the offset value

Reply:
Another suggestion: instead of giving feedback in bits and pieces like this, please review all the changes and make suggestions on the entire patchset. this small back and forth on little changes like this is annoying and provides lot of hindrance to contribution (assuming systemd community wants contributions from people outside the core group). cc: @poettering

Reply:
I guess it doesn't matter too much, you can ignore this comment.

Reply:
> Another suggestion: instead of giving feedback in bits and pieces like this, please review all the changes and make suggestions on the entire patchset.

The only thing that's left is https://github.com/systemd/systemd/pull/35281#discussion_r1864938466, everything else looks good.

Reply:
That I will argue is orthogonal to this work. For one thing, `test-chid-match.c` needs to be adjusted  to take into account the new devicetype parameter for `chid_match()`. Can we address that as a part of a separate PR?

Reply:
I'd argue that this needs to be done before merging this PR, because otherwise matching might lead to unexpected results if devices of both types are present in one table.

Reply:
Ok please suggest how to fix that test with exact code. I don't want to go back and forth readjusting the test code. I don't write it. The person who wrote is probably the best person to suggest how to rework that code.

Reply:
Sure. Obviously, there's more than one way to change that test. For example, this:
```
static struct {
        const char *name;
        const char *compatible;
        uint32_t type;
} results[] = {
        { "Device 1", "test,device-1", DEVICE_TYPE_DEVICETREE },
        { "Device 2", "test,device-2", DEVICE_TYPE_DEVICETREE },
        { "Device 3", "test,device-3", DEVICE_TYPE_DEVICETREE },
        { "Device 4", "test,vmware",   DEVICE_TYPE_FW },
};
```
And then pass `results[i].type` to `chid_match`. 

Do I also need to explain you how to make the `chid_match` changes? Those are fairly simple as well, just one `if` in the innermost cycle.